reconcile/terraform_cloudflare_users.py

```diff
@@ -0,0 +1,339 @@
+from reconcile.utils.external_resource_spec import ExternalResourceSpec
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import (
+    SecretReaderBase,
+    create_secret_reader,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.terraform import safe_resource_id
+from reconcile.utils.terraform.config_client import (
+    ClientAlreadyRegisteredError,
+    TerraformConfigClientCollection,
+)
+from reconcile.utils.terraform_client import run_terraform
+from reconcile.utils.terrascript.cloudflare_client import (
+    AccountShardingStrategy,
+    IntegrationUndefined,
+    InvalidTerraformState,
+    TerrascriptCloudflareClientFactory,
+)
+from reconcile.utils.terrascript.models import (
+    CloudflareAccount,
+    Integration,
+    TerraformStateS3,
+    VaultSecret,
+)
+
+QONTRACT_INTEGRATION = "terraform_cloudflare_users"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+QONTRACT_TF_PREFIX = "qrtfcfusers"
+CLOUDFLARE_EMAIL_DOMAIN_ALLOW_LIST_KEY = "cloudflareEmailDomainAllowList"
+
+
+@dataclass
+class CloudflareUser:
+    email_address: str
+    account_name: str
+    org_username: str
+    roles: Set[str]
+
+
+class TerraformCloudflareUsersParams(PydanticRunParams):
+    print_to_file: Optional[str]
+    account_name: Optional[str]
+    thread_pool_size: int
+    enable_deletion: bool
+
+
+class TerraformCloudflareUsers(
+    QontractReconcileIntegration[TerraformCloudflareUsersParams]
+):
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION.replace("_", "-")
+
+    def get_early_exit_desired_state(
+        self, *args: Any, **kwargs: Any
+    ) -> Optional[dict[str, Any]]:
+
+        cloudflare_roles, settings = self._get_desired_state()
+
+        return {
+            "cloudflare_roles": cloudflare_roles.dict(),
+            CLOUDFLARE_EMAIL_DOMAIN_ALLOW_LIST_KEY: settings.settings[0]
+            if settings.settings is not None
+            else [],
+        }
+
+    def _get_desired_state(
+        self,
+    ) -> Tuple[
+        CloudflareAccountRoleQueryData, AppInterfaceSettingCloudflareAndVaultQueryData
+    ]:
+        cloudflare_roles = terraform_cloudflare_roles.query(
+            query_func=gql.get_api().query
+        )
+
+        settings = app_interface_setting_cloudflare_and_vault.query(
+            query_func=gql.get_api().query
+        )
+        return cloudflare_roles, settings
+
+    def run(self, dry_run: bool) -> None:
+        print_to_file = self.params.print_to_file
+        account_name = self.params.account_name
+        thread_pool_size = self.params.thread_pool_size
+        enable_deletion = self.params.enable_deletion
+
+        cloudflare_roles, settings = self._get_desired_state()
+
+        secret_reader = create_secret_reader(
+            use_vault=settings.settings[0].vault
+            if settings.settings is not None
+            else True
```

Comment:
Not sure if we should raise an exception if settings are not found. That is currently the default behavior in other integrations, see https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/typed_queries/app_interface_vault_settings.py#L23-L30

Reply:
Yeah, good point. I missed this earlier. I'm raising the exception now.